general/mem_manager.hpp

```diff
@@ -68,148 +48,130 @@ class mm
-   /// corresponding device pointer (but not the provided host pointer).
+   /// Registers external host pointer in the memory manager which will manage
+   /// the corresponding device pointer, but not the provided host pointer.
    template<class T>
-   static inline void RegisterHostPtr(T *ptr_host, const size_t size)
+   void RegisterHostPtr(T *ptr_host, const std::size_t size)
    {
-      MM().Insert(ptr_host, size*sizeof(T));
+      Insert(ptr_host, size*sizeof(T));
 #ifdef MFEM_DEBUG
       RegisterCheck(ptr_host);
 #endif
    }
 
-   /// Registers external host and device pointers in the mm.
+   /// Registers external host and device pointers in the memory manager.
    template<class T>
-   static void RegisterHostAndDevicePtr(T *ptr_host, T *ptr_device,
-                                        const size_t size, bool host)
+   void RegisterHostAndDevicePtr(T *ptr_host, T *ptr_device,
+                                 const std::size_t size, const bool host)
    {
       RegisterHostPtr(ptr_host, size);
-      mm::memory &base = MM().maps.memories.at(ptr_host);
-      base.d_ptr = ptr_device;
-      base.host = host;
+      SetHostDevicePtr(ptr_host, ptr_device, host);
    }
 
-   /// Unregisters the host pointer from the mm. To be used with memory not
-   /// allocated by the mm.
-   template<class T>
-   static inline void UnregisterHostPtr(T *ptr)
-   {
-      if (!ptr) { return; }
-      mm::MM().Erase(ptr);
-   }
+   /// Set the host h_ptr, device d_ptr and mode host of the memory region just
+   /// been registered with h_ptr (see RegisterHostAndDevicePtr)
+   void SetHostDevicePtr(void *h_ptr, void *d_ptr, const bool host);
 
-   /// Check if pointer has been registered in the mm
-   static void RegisterCheck(void *ptr);
+   /// Unregisters the host pointer from the memory manager. To be used with
+   /// memory not allocated by the memory manager.
+   template<class T>
+   void UnregisterHostPtr(T *ptr) { Erase(ptr); }
 
-private:
-   ledger maps;
-   bool enabled;
-   mm(): enabled(true) { }
-   mm(mm const&) = delete;
-   void operator=(mm const&) = delete;
-   static inline mm& MM() { static mm singleton; return singleton; }
+   /// Check if pointer has been registered in the memory manager
+   void RegisterCheck(void *ptr);
 
-   /// Adds an address
-   void *Insert(void *ptr, const size_t bytes);
+   /// Prints all pointers known by the memory manager
+   void PrintPtrs(void);
 
-   /// Remove the address from the map, as well as all the address' aliases
-   void *Erase(void *ptr);
+   /// Copies all memory to the current memory space
+   void GetAll(void);
+};
 
-   /// Turn an address to the right host or device one
-   void *Ptr(void *ptr);
-   const void *Ptr(const void *ptr);
+/// The (single) global memory manager object
+extern MemoryManager mm;
 
-   /// Tests if ptr is a known address
-   bool Known(const void *ptr);
+/// Is global memory management still available?
+extern bool mm_destroyed;
```

Comment:
I don't have the whole picture, but maybe it would be nicer if this was a member of MemoryManager? The check could perhaps be moved into mm.Erase().

Reply:
We want a check to see if the mm hasn't been destroyed: that is why it is outside of the class and the check can't be in a mn.* function if the mm isn't there anymore. This is needed for example 9, that use a Vector outside of main, which can come to Delete itself after the mm was destroyed. Tested with Valgrind, this allows to get rid of the last leak, which was the mm object itself.

Reply:
Could it be a static member? If the check was the first thing in Erase, it should work even if the instance is gone. It would be practically the same thing, we would just get rid of the suspicious looking global.

Reply:
I guess so, sure: that would be better. We can change that.

Reply:
done